Enhance song recommendation system with new features and dependencies (*first draft*)

claiireyu opened the pull request:
- Added Flyway database migration dependencies in pom.xml.
- Updated song recommendation requirements to include genre and artist distribution analysis, configurable recommendation limits, and a detailed scoring system.
- Implemented HealthController for database connection health check.
- Expanded SongController to handle recommendations, genre distribution, and artist distribution endpoints.
- Introduced RecommendationService for processing song recommendations based on user input.
- Updated Recommendation and Song models to include similarity scores and improved data handling.
- Enhanced application properties for database and Flyway configurations, along with JSON formatting options.

claiireyu committed:
Add unit testing workflow and update README

claiireyu committed:
Update GitHub Actions workflow to set up JDK 21 for Maven tests

claiireyu committed:
Add GitHub Actions workflow for production deployment with Docker and AWS integration

claiireyu committed:
Fix AWS credentials and update service name in GitHub Actions deployment workflow

claiireyu committed:
Refactor GitHub Actions workflow to clarify Docker image build step

claiireyu committed:
Update GitHub Actions workflow to enhance clarity in Docker image build step

claiireyu committed:
commit all local changes

claiireyu committed:
Merge remote-tracking branch 'origin/main' into deploy-ecs

claiireyu committed:
Merge branch 'main' into deploy-ecs

claiireyu committed:
Fix Docker image build step in GitHub Actions workflow by correcting ECR repository variable usage

claiireyu committed:
Enhance GitHub Actions workflow by adding output for Docker image tag after build and push

claiireyu committed:
Add AWS ECS task definition for song recommendation service

claiireyu committed:
Remove AWS ECS task definition for song recommendation service

claiireyu committed:
Add execution role ARN to AWS ECS task definition for song recommendation service

claiireyu committed:
Add Spring Boot Actuator dependency and configure health endpoint in application properties

claiireyu committed:
Add health check configuration to AWS ECS task definition for song recommendation service

claiireyu committed:
Add initial requirements document for Song Recommendation System, outlining phases, input/output specifications, feature weights, scoring algorithm, and technical notes.

claiireyu committed:
Enhance song recommendation system with new features and dependencies (*first draft*)

- Added Flyway database migration dependencies in pom.xml.
- Updated song recommendation requirements to include genre and artist distribution analysis, configurable recommendation limits, and a detailed scoring system.
- Implemented HealthController for database connection health check.
- Expanded SongController to handle recommendations, genre distribution, and artist distribution endpoints.
- Introduced RecommendationService for processing song recommendations based on user input.
- Updated Recommendation and Song models to include similarity scores and improved data handling.
- Enhanced application properties for database and Flyway configurations, along with JSON formatting options.

claiireyu committed:
Merge branch 'main' into rec-phase-1

claiireyu committed:
Update test configurations and dependencies for improved testing environment

- Added H2 and Spring Boot Test Starter dependencies in pom.xml for in-memory database testing.
- Refactored application.properties for test profile to use H2 database and disable Flyway migrations.
- Removed unused TestConfig class and its associated DataSource bean.
- Updated SongRecommendationApplicationTests to simplify test context configuration.

These changes enhance the testing setup, ensuring a more efficient and isolated testing environment.

claiireyu committed:
Merge branch 'rec-phase-1' of https://github.com/claiireyu/spotify-recommendation into rec-phase-1
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

johnyu617 commented on requirements/song_rec.md:
```diff
@@ -0,0 +1,72 @@
+# Song Recommendation System Requirements
+
+## Phase 1: Basic Recommendation System
```
Is this PR implement phase 1. If so, please update the PR description

johnyu617 commented on src/test/java/com/example/demo/config/TestConfig.java:
```diff
@@ -2,18 +2,10 @@
-            .setType(EmbeddedDatabaseType.H2)
-            .build();
-    }
+    // Add any test-specific beans here if needed
```
could you add unit tests? You can submit a different PR.
3. Write Unit Tests for Controllers
	•	Use @WebMvcTest to load only the web layer components.
	•	Use MockMvc to test HTTP requests and responses without starting the server.

4. Write Unit Tests for Service Layer
	•	Test the service logic in isolation using mock objects for repositories or external services.
	•	Use Mockito for mocking dependencies.

5. Write Unit Tests for Repository Layer
	•	Use @DataJpaTest for repository tests, which loads an in-memory database like H2.

johnyu617 commented on src/main/java/com/example/demo/service/RecommendationService.java:
```diff
@@ -0,0 +1,154 @@
+        logger.info("Frequent artists: {}", frequentArtists);
+
+        // Get all available songs from database
+        List<Song> availableSongs = songRepository.findAll();
```
Spotify’s music library has grown to over 100 million tracks. load all songs to the database will cause performance issue.

claiireyu committed:
Add unit tests for SongController to validate recommendation features
- Ensured proper handling of edge cases such as null inputs, empty song lists, and service exceptions.
- Utilized MockMvc for testing the controller endpoints, enhancing the reliability of the song recommendation system.
